homeassistant/components/hko/__init__.py

```diff
@@ -0,0 +1,260 @@
+    district = next(
+        (item for item in LOCATIONS if item[KEY_LOCATION] == location),
+        {KEY_DISTRICT: DEFAULT_DISTRICT},
+    )[KEY_DISTRICT]
+    websession = async_get_clientsession(hass)
+
+    coordinator = HKOUpdateCoordinator(hass, websession, district, location)
+    await coordinator.async_config_entry_first_refresh()
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class HKOUpdateCoordinator(DataUpdateCoordinator):
+    """HKO Update Coordinator."""
+
+    def __init__(self, hass: HomeAssistant, session, district, location) -> None:
+        """Update data via library."""
+        self.location = location
+        self.district = district
+        self.hko = HKO(session)
+
+        update_interval = timedelta(minutes=10)
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            # update_method=self._async_update_data(),
+            update_interval=update_interval,
+        )
+
+    async def _async_update_data(self):
+        """Update data via HKO library."""
+        try:
+            async with timeout(60):
+                rhrread = await self.hko.weather("rhrread")
+                fnd = await self.hko.weather("fnd")
+        except HKOError as error:
+            raise UpdateFailed(error) from error
+        return {
+            API_CURRENT: self._convert_current(rhrread),
+            API_FORECAST: [
+                self._convert_forecast(item) for item in fnd[API_WEATHER_FORECAST]
+            ],
+        }
+
+    def _convert_current(self, data):
+        current = {
+            API_HUMIDITY: data[API_HUMIDITY][API_DATA][0][API_VALUE],
+            API_TEMPERATURE: next(
+                (
+                    item[API_VALUE]
+                    for item in data[API_TEMPERATURE][API_DATA]
+                    if item[API_PLACE] == self.location
+                ),
+                0,
+            ),
+        }
+        return current
+
+    def _convert_forecast(self, data):
+        date = data[API_FORECAST_DATE]
+        forecast = {
+            ATTR_FORECAST_CONDITION: self._convert_icon_condition(
+                data[API_FORECAST_ICON], data[API_FORECAST_WEATHER]
+            ),
+            ATTR_FORECAST_TEMP: data[API_FORECAST_MAX_TEMP][API_VALUE],
+            ATTR_FORECAST_TEMP_LOW: data[API_FORECAST_MIN_TEMP][API_VALUE],
+            ATTR_FORECAST_TIME: f"{date[0:4]}-{date[4:6]}-{date[6:8]}T00:00:00+08:00",
+        }
+        return forecast
+
+    def _convert_icon_condition(self, icon, info):
```

Comment:
In my opinion you should use a dictionary for all the icons instead of multiple return statements.

Reply:
That's a good point. However, some icons do not have a suitable translation to a weather condition in Home Assistant, so my idea is that the return statements act as a fallback. Admittedly it is pretty messy, do you think there is a better way of doing it?

Reply:
Moved icon mapping to const